### PR TITLE
Adding matmul.

### PR DESCRIPTION
**Pull request by Narsil**

### CODE REVIEW

File: src/tensor.rs [hunk 1]
```diff
@@ -233,6 +253,57 @@ impl Tensor {
         Ok(Self(Arc::new(tensor_)))
     }
 
+    pub fn matmul(&self, rhs: &Self) -> Result<Self> {
+        let a_dims = self.shape().dims();
+        let b_dims = rhs.shape().dims();
+
+        let dim = a_dims.len();
+
+        // if dim < 2 {
```
Comment (LaurentMazare) [inline src/tensor.rs:262]: Re-enable this test as part of the `if b_dims.len() < 2`, i.e. `if dim < 2 || b_dims.len() < 2`?

File: src/cpu_backend.rs [hunk 2]
```diff
@@ -97,6 +106,38 @@ impl CpuStorage {
...
+        println!("lhs_stride {lhs_stride:?}");
+        println!("rhs_stride {rhs_stride:?}");
+        // todo!("matmul");
+        let a_skip: usize = m * k;
+        let b_skip: usize = n * k;
+        let c_skip: usize = m * n;
+
+        let mut c = Self::F32(vec![0.0; b * m * n]);
+
+        batched_sgemm(
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/cpu_backend.rs]: Did simple sanity checks for now

File: src/cpu_backend.rs [hunk 3]
```diff
@@ -97,6 +106,38 @@ impl CpuStorage {
...
     }
 
+    pub(crate) fn matmul_impl(
+        &self,
+        rhs: &Self,
+        (b, m, n, k): (usize, usize, usize, usize),
+        lhs_stride: &[usize],
+        rhs_stride: &[usize],
+    ) -> Result<Self> {
+        println!("rhs {rhs:?}");
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/cpu_backend.rs]: Removed all of these, sorry forgot to push my cleanup yesterday :)

**Comment by Narsil** [comment]
Should be ready for real review now.

File: src/tensor.rs [hunk 4]
```diff
@@ -147,11 +147,12 @@ impl Tensor {
 
     pub fn new_impl<A: crate::device::NdArray>(
         array: A,
+        shape: Shape,
         device: &Device,
         is_variable: bool,
     ) -> Result<Self> {
-        let shape = array.shape()?;
-        let storage = device.tensor(array)?;
+        // let shape = array.shape()?;
```
Comment (LaurentMazare) [inline src/tensor.rs:154]: Remove this comment?

File: src/tensor.rs [hunk 5]
```diff
@@ -234,10 +253,65 @@ impl Tensor {
...
+            return Err(Error::ShapeMismatchBinaryOp {
+                lhs: self.shape().clone(),
+                rhs: rhs.shape().clone(),
+                op: "matmul",
+            });
+        }
+
+        let mut c_shape: Vec<_> = a_dims[..dim - 2].into();
+        c_shape.extend(&[m, n]);
+        let c_shape: Shape = Shape(c_shape);
```
Comment (LaurentMazare) [inline src/tensor.rs:284]: Probably no need for the type annotation here? (really cosmetic, it just makes me wonder why it's here)

**Review by LaurentMazare — APPROVED** [review]

File: src/tensor.rs [hunk 6]
```diff
@@ -151,7 +151,11 @@ impl Tensor {
         device: &Device,
         is_variable: bool,
     ) -> Result<Self> {
-        // let shape = array.shape()?;
+        let n: usize = shape.0.iter().product();
+        let buffer_size: usize = array.shape()?.0.iter().product();
```
Comment (LaurentMazare) [inline src/tensor.rs:155]: `array.shape()?.element_count()`?

Comment (Narsil) [reply src/tensor.rs:155]: `elem_count` !

File: src/tensor.rs (hunk 5 above)
Comment (Narsil) [reply src/tensor.rs]: I'm shadowing the Vec, so I sometimes go over the top that this is indeed shadowing.

**Review by LaurentMazare — APPROVED** [review]

**Pull request merged**